### ISSUE TITLE
Enhance Notification Settings Logic for User Entity Updates in CouchDB

### ISSUE THREAD

**Issue by Ayush8923** [Aam-Digital/ndb-core#2723]
**Objective**
This issue involves refining the notification settings logic and ensuring seamless synchronization with CouchDB. The following tasks are outlined:

**Proposed Solutions & Alternatives**
- [x] **Update Notification Logic:**
    - Implement logic to allow users to select any entity and define the corresponding notification conditions. These conditions should be saved and updated in the `User` entity in CouchDB.

---

- [x] **Database Schema Example**
Below is an example of the expected schema for a `NotificationConfig` entity after notification settings are applied:

```json
{
    "_id": "NotificationConfig:<user-uuid>",
    "_rev": "2-f9e14ee88f8107bd3a7bd5ec82f40fa4",
    "created": {
        "at": "2024-01-08T17:45:16.529Z",
        "by": "User:demo"
    },
    "updated": {
        "at": "2024-01-08T17:45:16.529Z",
        "by": "User:demo"
    },
    "notificationTypes": [
        {
            "entityType": "User",
            "conditions": {
                "field": "name",
                "operator": "eq",
                "value": "Anilaabh"
            },
            "isNotificationEnabled": true
        },
        {
            "entityType": "Task",
            "conditions": {
                "field": "name",
                "operator": "eq",
                "value": "Anilaabh"
            },
            "isNotificationEnabled": true
        }
    ]
}
```

---

- [x] **Push Notification Toggle Logic:**
    - Ensure that when the user enables or disables push notifications from the "Where you receive notifications" section in the notification settings page, the notificationsEnabled field in CouchDB's `User` entity is updated accordingly.

- [x] **Add/Delete Notification Rules:**
    - Implement logic to handle the addition and deletion of notification rules. When rules are added or removed, ensure the data is consistently updated in the `User` entity in CouchDB.
- [x] The Save button should be enabled only if any changes are made to the notification settings.

- [x] **Retrieve and Update UI:**
    - Fetch all notification rules associated with the user and update the UI dynamically to reflect these rules.
    
- [ ] Update the Notification rule condition UI(maybe need to add the operator field) and based on that user can set the condition accordingly. https://github.com/Aam-Digital/ndb-core/issues/2775

**Comment by sleidig** [comment]
Just one remark here, @Ayush8923 : The configuration should be stored as its own Entity (maybe "NotificationConfig:<user_uuid>" or whatever name you suggest as more fitting).
We recently have move away from combining any different data into one entity for various reasons and not every customized system even has a "User" entity type now.

**Comment by Ayush8923** [comment]
> Just one remark here, @Ayush8923 : The configuration should be stored as its own EntityType (maybe "NotificationConfig:<user_uuid>" or whatever name you suggest as more fitting). We recently have moved away from combining any different data into one entity for various reasons and not every customized system even has a "User" entity type now.

Thanks for this remark @sleidig. I have updated the draft database schema in the issue description. Please take a look once and let me know if there are any changes needed.

**Comment by sleidig** [comment]
as per group discussion:
```json
{
    "_id": "NotificationConfig:<userId (Keycloak ID)>",
    "_rev": ...,
    "created": ...,
    "updated": ...,
    "channels": { push: true, email: false }, 
    "notificationTypes": [
        {
            "notificationType": "entity-update" | "new_app_release",
            "enabled": true,
            "channels": { push: true, email: false },
            "entityType": "User",
            "conditions" // as <DataFilter/MongoQuery>, see https://www.mongodb.com/docs/manual/reference/operator/query/
        },
        {
            "entityType": "Task",
            "conditions": {
                "authors": { "$in": "<userId>" },
            }
        }
    ]
}
```

**Comment by Ayush8923** [comment]
@sleidig @tomwwinter @george-neha, In the notification setting page, I have created an initial UI(attached screenshot below) for the notification condition field. In this UI, the user first selects an entity type and then the user adds the corresponding fields (e.g., subject, name, etc.) with which the user wants to perform the comparison. Then the user selects an operator (e.g., Equals, Greater than, etc.) and specifies a value for the comparison in the notification condition.
Please review this initial UI and share your inputs or suggestions. Also, please let me know if I have missed anything.

<img width="1504" alt="Screenshot 2024-12-27 at 7 19 26 PM" src="https://github.com/user-attachments/assets/04281b00-646f-442d-834f-704deb3718c6" />

**Comment by sleidig** [comment]
We should support multiple conditions (so probably similar to the "add new notification rule" button, we can have a "add condition" button to add another block of this). Accordingly, I would move the condition fields to a second row in the box and maybe make that expandable/collapsible for easier overview of all rules.

A general thought coming to my mind: These conditions are actually similar (or probably the same?) as the filters to filter entries in the EntityList views. Can we reuse those UI components to filter data for the notification conditions instead of introducing a new model?

**Comment by Ayush8923** [comment]
> We should support multiple conditions (so probably similar to the "add new notification rule" button, we can have a "add condition" button to add another block of this). Accordingly, I would move the condition fields to a second row in the box and maybe make that expandable/collapsible for easier overview of all rules.
> 
> A general thought coming to my mind: These conditions are actually similar (or probably the same?) as the filters to filter entries in the EntityList views. Can we reuse those UI components to filter data for the notification conditions instead of introducing a new model?

Yes, I agree with the idea of supporting multiple conditions. Implementing an `Add Condition` button (maybe we can add the plus icon) for each rule will make the feature more flexible.

Yes, Instead of creating a new component, we can use the existing EntityList filter components for notification conditions. I’ll check the current filter implementation to evaluate how it can be adapted or extended for this use case.
